public/app/plugins/datasource/stackdriver/components/Aggregations.tsx

```diff
@@ -18,81 +18,65 @@ export interface Props {
+                <Icon name={displayAdvancedOptions ? 'angle-down' : 'angle-right'} /> Advanced Options
+              </>
+            </a>
+          </label>
+        </div>
+      </div>
+      {props.children(displayAdvancedOptions)}
+    </>
+  );
+};
 
-  setAggOptions({ metricDescriptor }: Props) {
-    let aggOptions: any[] = [];
-    if (metricDescriptor) {
-      aggOptions = getAggregationOptionsByMetric(
-        metricDescriptor.valueType as ValueTypes,
-        metricDescriptor.metricKind as MetricKind
-      ).map(a => ({
-        ...a,
-        label: a.text,
-      }));
+const useAggregationOptionsByMetric = ({ metricDescriptor }: Props): Array<SelectableValue<string>> => {
+  return useMemo(() => {
+    if (!metricDescriptor) {
+      return [];
     }
-    this.setState({ aggOptions });
-  }
 
-  onToggleDisplayAdvanced = () => {
-    this.setState(state => ({
-      displayAdvancedOptions: !state.displayAdvancedOptions,
+    return getAggregationOptionsByMetric(
+      metricDescriptor.valueType as ValueTypes,
+      metricDescriptor.metricKind as MetricKind
+    ).map(a => ({
+      ...a,
+      label: a.text,
     }));
-  };
-
-  render() {
-    const { displayAdvancedOptions, aggOptions } = this.state;
-    const { templateVariableOptions, onChange, crossSeriesReducer } = this.props;
+  }, [metricDescriptor?.metricKind, metricDescriptor?.valueType]);
+};
 
-    return (
-      <>
-        <div className="gf-form-inline">
-          <label className="gf-form-label query-keyword width-9">Aggregation</label>
-          <Segment
-            onChange={({ value }) => onChange(value)}
-            value={[...aggOptions, ...templateVariableOptions].find(s => s.value === crossSeriesReducer)}
-            options={[
-              {
-                label: 'Template Variables',
-                options: templateVariableOptions,
-              },
-              {
-                label: 'Aggregations',
-                expanded: true,
-                options: aggOptions,
-              },
-            ]}
-            placeholder="Select Reducer"
-          ></Segment>
-          <div className="gf-form gf-form--grow">
-            <label className="gf-form-label gf-form-label--grow">
-              <a onClick={this.onToggleDisplayAdvanced}>
-                <>
-                  <Icon name={displayAdvancedOptions ? 'angle-down' : 'angle-right'} /> Advanced Options
-                </>
-              </a>
-            </label>
-          </div>
-        </div>
-        {this.props.children && this.props.children(this.state.displayAdvancedOptions)}
-      </>
-    );
-  }
-}
+const useSelectedFromOptions = (aggOptions: Array<SelectableValue<string>>, props: Props) => {
+  return useMemo(() => {
+    const allOptions = [...aggOptions, ...props.templateVariableOptions];
+    return allOptions.find(s => s.value === props.crossSeriesReducer);
+  }, [aggOptions, props.crossSeriesReducer, props.templateVariableOptions]);
+};
```

Comment:
@mckn this and https://github.com/grafana/grafana/pull/21291/files#diff-23a463c53eb7a50138db255fed80260cR61 are quite similar - can we have a common reusable hook,  something like this crossed my head: `useSelectedFromOptions<T>(predicate: () => SelectableValue<T>, dependencies?)` ?

Reply:
yes, that sounds like a brilliant idea! I will have a look at that.

Reply:
Tho, it will not really fit in this scenario because of the need of filtering two combined lists. So I will skip this one for this PR and lets open a new one where we introduce the hook you suggested.